File: C#.Net/JDI/Web/JDI Web/Selenium/Elements/Complex/BaseSelector.cs
using System;
using System.Collections.Generic;
using System.Linq;
using JDI_Commons;
using Epam.JDI.Core.Interfaces.Base;
using JDI_Web.Selenium.Base;
using JDI_Web.Selenium.DriverFactory;
using JDI_Web.Selenium.Elements.Base;
using JDI_Web.Settings;
using OpenQA.Selenium;
using OpenQA.Selenium.Support.UI;
using static Epam.JDI.Core.Settings.JDISettings;

namespace JDI_Web.Selenium.Elements.Complex
{
    public abstract class BaseSelector<TEnum> : WebBaseElement, IVisible
        where TEnum : IConvertible
    {
        protected bool IsSelector;
        protected GetElementType _allLabels;
        protected SelectElement Selector
        {
            get
            {
                IsSelector = true;
                return new SelectElement(WebElement);
            }
        }

        protected BaseSelector(By optionsNamesLocator, List<IWebElement> webElements = null, WebBaseElement element = null) : 
            base(optionsNamesLocator, webElements: webElements, element:element) { }

        protected BaseSelector(By optionsNamesLocator, By allLabelsLocator, List<IWebElement> webElements = null, WebBaseElement element = null) 
            : base(optionsNamesLocator, webElements:webElements, element:element)
        {
            _allLabels = new GetElementType(allLabelsLocator, this);
        }

        public TextList AllLabels => _allLabels?.Get(new TextList(), WebAvatar);

        public Action<BaseSelector<TEnum>, string> SelectNameAction = (s, name) =>
        {
            if (!s.HasLocator && s.AllLabels == null)
                throw Exception($"Can't find option '{name}'. No optionsNamesLocator and _allLabelsLocator found");
            if (s.Locator.ToString().Contains("{0}"))
            {
                new Clickable(s.Locator.FillByTemplate(name), element:s).Click();
                return;
            }
            if (s.AllLabels != null)
            {
                s.SelectFromList(s.AllLabels.WebElements, name);
                return;
            }
            var elements = s.WebAvatar.SearchAll().WebElements;
            var selector = GetSelectElement(elements);
            if (selector != null)
                if (selector.Options.Any())
                {
                    selector.SelectByText(name);
                    return;
                }
                else                    
                    throw Exception($"<select> tag has no <option> tags. Please Clarify element locator ({s})");
            if (elements.Count == 1 && elements[0].TagName.Equals("ul"))
                elements = elements[0].FindElements(By.TagName("li")).ToList();
            s.SelectFromList(elements, name);
        };

        private static SelectElement GetSelectElement(List<IWebElement> elements)
        {
            var selector = elements.Count == 1 ? elements[0] : null;
            if (selector == null 
                && elements.Count(el => WebSettings.WebDriverFactory.ElementSearchCriteria(el) 
                && el.TagName.Equals("select")) == 1)
                selector = elements
                    .First(el => WebSettings.WebDriverFactory.ElementSearchCriteria(el) && el.TagName.Equals("select"));
            return selector != null ? new SelectElement(selector) : null;
        }

        private void SelectFromList(IList<IWebElement> els, string name)
        {
            var element = els.FirstOrDefault(el => el.Text.Equals(name));
            if (element == null)
                throw Exception($"Can't find option '{name}'. Please fix _allLabelsLocator");
            element.Click();
        }

        public Action<BaseSelector<TEnum>, int> SelectNumAction = (s, num) =>
        {
            if (!s.HasLocator && s.AllLabels == null)
                throw Exception($"Can't find option '{num}'. No optionsNamesLocator and _allLabelsLocator found");
            if (s.AllLabels != null)
            {
                s.SelectFromList(s.AllLabels.WebElements, num);
                return;
            }
            if (s.Locator.ToString().Contains("{0}"))
            {
                new Clickable(s.Locator.FillByTemplate(num)).Click();
                return;
            }
            var elements = s.WebAvatar.SearchAll().WebElements;
            if (elements.Count == 1 && elements[0].TagName.Equals("select"))
                if (s.Selector.Options.Any())
                {
                    s.Selector.SelectByIndex(num - 1);
                    return;
                }
                else
                    throw Exception($"<select> tag has no <option> tags. Please Clarify element locator ({s})");
            if (elements.Count == 1 && elements[0].TagName.Equals("ul"))
                elements = elements[0].FindElements(By.TagName("li")).ToList();
            s.SelectFromList(elements, num);
        };

        private void SelectFromList(IList<IWebElement> els, int num)
        {
            if (num <= 0)
                throw Exception($"Can't get option with num '{num}'. num should be 1 or more");
            if (els == null)
                throw Exception($"Can't find option with num '{num}'. Please fix _allLabelsLocator");
            if (els.Count < num)
                throw Exception($"Can't find option with num '{num}'. Find only '{els.Count}' options");
            els[num - 1].Click();
        }

        public Func<BaseSelector<TEnum>, string, bool> SelectedNameAction;
        public Func<BaseSelector<TEnum>, int, bool> SelectedNumAction;
        public Func<BaseSelector<TEnum>, string> GetValueAction;

        public Func<BaseSelector<TEnum>, IWebElement, bool> SelectedElementAction = (s, el) =>
        {
            if (s.IsSelector)
                return el.Selected;
            var attr = el.GetAttribute("checked");
            return attr != null && attr.Equals("true");
        };

        public void WaitSelected(string name)
        {
            Actions.Selected(name, (el, n) => TimerExtensions.ForceDone(() => SelectedNameAction(this, n)));
        }

        public void WaitSelected(TEnum enumType)
        {
            WaitSelected(enumType.ToString());
        }

        public bool Selected(string name)
        {
            return Actions.Selected(name, (el, n) => SelectedNameAction(this, n));
        }

        public bool Selected(TEnum enumType)
        {
            return Selected(enumType.ToString());
        }

        public Func<BaseSelector<TEnum>, List<string>> GetOptionsAction = 
            s => s.Elements.Select(el => el.Text).ToList();


        public virtual Action<BaseSelector<TEnum>, string> SetValueAction { get; set; } = (s, value) => s.SelectNameAction(s, value);

        public string Value
        {
            get { return Actions.GetValue(el => GetValueAction(this)); }
            set { Actions.SetValue(value, (el, val) => SetValueAction(this, val)); }
        }

        public IList<string> Options => GetOptionsAction(this);
        public IList<string> Names => Options;
        public IList<string> Values => Options;

        public string OptionsAsText => Options.Print();

        public IList<IWebElement> Elements
        {
            get
            {
                if (!HasLocator && AllLabels == null)
                    throw Exception(
                        "Can't check is element displayed or not. No optionsNamesLocator and allLabelsLocator found");
                if (AllLabels != null)
                    return AllLabels.WebElements;
                if (Locator.ToString().Contains("{0}"))
                    throw Exception(
                        "Can't check is element displayed or not. Please specify allLabelsLocator or correct optionsNamesLocator (should not contain '{0}')");
                return GetElementsFromTag();
            }
        }

        public IList<IWebElement> GetElementsFromTag()
        {
            IList<IWebElement> elements;
            try { elements = WebAvatar.SearchAll().WebElements; }
            catch { return new List<IWebElement>(); }
            if (elements.Count == 1)
                switch (elements[0].TagName)
                {
                    case "select":
                        return Selector.Options;
                    case "ul":
                        return elements[0].FindElements(By.TagName("li"));
                }
            return elements;
        }

        public Func<BaseSelector<TEnum>, string, IWebElement> GetWebElementFunc = (s, name) =>
        {
            if (!s.HasLocator)
                throw Exception("Element has no locators");
            return s.Locator.ToString().Contains("{0}")
                ? new WebElement(s.Locator.FillByTemplate(name))
                {
                    WebAvatar = { DriverName = s.WebAvatar.DriverName },
                    Parent = s.Parent
                }.WebElement
                : s.Elements.FirstOrDefault(el => el.Text.Equals(name));
        };
        public IWebElement GetWebElement(string name)
        {
            return GetWebElementFunc(this, name);
        }

        public Func<BaseSelector<TEnum>, string, bool> DisplayedNameAction = (s, name) =>
        {
            var el = s.GetWebElement(name);
            return el != null && el.Displayed;
        };

        public Func<BaseSelector<TEnum>, int, bool> DisplayedNumAction = 
            (s, num) => s.DisplayedInList(s.Elements, num);

        private bool DisplayedInList(IList<IWebElement> els, int num)
        {
            if (num <= 0)
                throw Exception($"Can't get option with num '{num}'. num should be 1 or more");
            if (els == null)
                throw Exception($"Can't find option with num '{num}'. Please fix _allLabelsLocator");
            if (els.Count < num)
                throw Exception($"Can't find option with num '{num}'. Find '{els.Count}' options");
            return els[num - 1].Displayed;
        }

        public Func<BaseSelector<TEnum>, bool> DisplayedAction = s =>
        {
            var els = s.WebAvatar.FindImmediately(() => s.Elements, null);
            return els != null && els.Any() && els[0].Displayed;
        };

        public Func<BaseSelector<TEnum>, bool> WaitDisplayedAction = s => 
        {
            return s.Timer.Wait(() => {
                var els = s.Elements;
                return els != null && els.Any() && els[0].Displayed;
            });
        };

        public Func<BaseSelector<TEnum>, bool> WaitVanishedAction = 
            s => s.Timer.Wait(() => !s.DisplayedAction(s));

        public bool Displayed => Actions.IsDisplayed(s => DisplayedAction(this));

        public bool Hidden => Actions.IsDisplayed(s => !DisplayedAction(this));

        public void WaitDisplayed()
        {
            Actions.WaitDisplayed(s => WaitDisplayedAction(this));
        }

        public void WaitVanished()
        {
            Actions.WaitVanished(s => Timer.Wait(() => !DisplayedAction(this)));
        }
    }
}

File: Java/JDI/jdi-uitest-web/src/main/java/com/epam/jdi/uitests/web/selenium/elements/complex/BaseSelector.java
package com.epam.jdi.uitests.web.selenium.elements.complex;
/*
 * Copyright 2004-2016 EPAM Systems
 *
 * This file is part of JDI project.
 *
 * JDI is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * JDI is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; 
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with JDI. If not, see <http://www.gnu.org/licenses/>.
 */


import com.epam.jdi.uitests.core.interfaces.base.IVisible;
import com.epam.jdi.uitests.web.selenium.elements.GetElementType;
import com.epam.jdi.uitests.web.selenium.elements.base.BaseElement;
import com.epam.jdi.uitests.web.selenium.elements.base.Clickable;
import com.epam.jdi.uitests.web.selenium.elements.base.Element;
import com.epam.jdi.uitests.web.selenium.elements.base.IHasElement;
import org.openqa.selenium.By;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.support.ui.Select;

import java.util.ArrayList;
import java.util.List;

import static com.epam.commons.EnumUtils.getEnumValue;
import static com.epam.commons.LinqUtils.first;
import static com.epam.commons.LinqUtils.select;
import static com.epam.commons.ReflectionUtils.isInterface;
import static com.epam.commons.Timer.waitCondition;
import static com.epam.jdi.uitests.core.settings.JDISettings.exception;
import static com.epam.jdi.uitests.web.selenium.driver.WebDriverByUtils.fillByTemplate;

/**
 * Created by Roman_Iovlev on 7/9/2015.
 */

abstract class BaseSelector<TEnum extends Enum> extends BaseElement implements IVisible, IHasElement {
    protected boolean isSelector;
    protected GetElementType allLabels = new GetElementType();

    BaseSelector() {
        super();
    }

    BaseSelector(By optionsNamesLocator) {
        super(optionsNamesLocator);
    }

    BaseSelector(By optionsNamesLocator, By allLabelsLocator) {
        super(optionsNamesLocator);
        this.allLabels = new GetElementType(allLabelsLocator, this);
    }

    protected TextList allLabels() {
        return allLabels.get(TextList.class);
    }

    protected void selectAction(String name) {
        Clickable cl = getElement(name);
        if (cl != null)
            cl.click();
    }
    protected Clickable getElement(String name) {
        if (!hasLocator() && allLabels() == null)
            throw exception("Can't find option '%s'. No optionsNamesLocator and allLabelsLocator found", name);
        if (hasLocator() && getLocator().toString().contains("%s"))
            return new GetElementType(fillByTemplate(getLocator(), name), this).get(Clickable.class);
        if (allLabels() != null)
            return allLabels().getLocator().toString().contains("%s")
                ? new GetElementType(fillByTemplate(allLabels().getLocator(), name), this).get(Clickable.class)
                : getFromList(allLabels().avatar.searchAll().getElements(), name);
        List<WebElement> elements = getAvatar().searchAll().getElements();
        WebElement element = elements.get(0);
        if (elements.size() == 1 && element.getTagName().equals("select"))
            if (getSelector().getOptions().size() > 0) {
                getSelector().selectByVisibleText(name);
                return null;
            }
            else throw exception("<select> tag has no <option> tags. Please Clarify element locator (%s)", this);
        if (elements.size() == 1 && element.getTagName().equals("ul"))
            elements = element.findElements(By.tagName("li"));
        return getFromList(elements, name);
    }

    private Clickable getFromList(List<WebElement> els, String name) {
        WebElement element = first(els, el -> el.getText().toLowerCase().trim().equals(name.toLowerCase().trim()));
        if (element == null)
            throw exception("Can't find option '%s'. Please fix allLabelsLocator", name);
        Clickable cl = new Clickable(element);
        cl.setParent(getParent());
        return cl;
    }
    public WebElement getWebElement() {
        if (avatar.hasWebElement())
            return avatar.getElement();
        Object parent = getParent();
        return parent != null && isInterface(parent.getClass(), IHasElement.class)
                ? ((IHasElement)parent).getWebElement()
                : null;
    }
    public void setWebElement(WebElement webElement) {
        Element element = new Element();
        element.setParent(getParent());
        element.setWebElement(webElement);
        setParent(element);
    }

    protected void selectAction(int num) {
        Clickable cl = getElement(num);
        if (cl != null)
            cl.click();
    }
    protected Clickable getElement(int num) {
        if (!hasLocator() && allLabels() == null)
            throw exception("Can't find option '%s'. No optionsNamesLocator and allLabelsLocator found", num);
        if (allLabels() != null)
            return selectFromList(allLabels().avatar.searchAll().getElements(), num);
        if (getLocator().toString().contains("%s"))
            return new GetElementType(fillByTemplate(getLocator(), num), this).get(Clickable.class);
        List<WebElement> elements = getAvatar().searchAll().getElements();
        WebElement element = elements.get(0);
        if (elements.size() == 1 && element.getTagName().equals("select"))
            if (getSelector().getOptions().size() > 0) {
                getSelector().selectByIndex(num - 1);
                return null;
            }
            else throw exception("<select> tag has no <option> tags. Please Clarify element locator (%s)", this);
        if (elements.size() == 1 && element.getTagName().equals("ul"))
            elements = element.findElements(By.tagName("li"));
        return selectFromList(elements, num);
    }

    private Clickable selectFromList(List<WebElement> els, int num) {
        if (num <= 0)
            throw exception("Can't get option with num '%s'. Number should be 1 or more", num);
        if (els == null)
            throw exception("Can't find option with num '%s'. Please fix allLabelsLocator", num);
        if (els.size() < num)
            throw exception("Can't find option with num '%s'. Find only '%s' options", num, els.size());
        return new Clickable(els.get(num - 1));
    }

    protected abstract boolean isSelectedAction(String name);

    protected abstract boolean isSelectedAction(int num);

    protected boolean isSelectedAction(WebElement el) {
        if (isSelector)
            return el.isSelected();
        String attr = el.getAttribute("checked");
        return attr != null && attr.equals("true");
    }

    /**
     * @param name Specify name using string
     * Wait while option (from text) is selected. Return false if this not happens
     */
    public final void waitSelected(String name) {
        actions.isSelected(name, n -> waitCondition(() -> isSelectedAction(n)));
    }

    /**
     * @param name Specify name using enum
     * Wait while option (from enum) is selected. Return false if this not happens
     */
    public final void waitSelected(TEnum name) {
        waitSelected(getEnumValue(name));
    }

    /**
     * @param name Specify name using string
     * @return Is option selected?
     */
    public final boolean isSelected(String name) {
        return actions.isSelected(name, this::isSelectedAction);
    }

    /**
     * @param name Specify name using enum
     * @return Is option selected?
     */
    public final boolean isSelected(TEnum name) {
        return isSelected(getEnumValue(name));
    }

    protected List<String> getOptionsAction() {
        return select(getElements(), WebElement::getText);
    }

    protected abstract String getValueAction();

    protected void setValueAction(String value) {
        selectAction(value);
    }

    /**
     * @return Get value of Element
     */
    public final String getValue() {
        return actions.getValue(this::getValueAction);
    }

    /**
     * @param value Specify element value
     *              Set value to Element
     */
    public final void setValue(String value) {
        actions.setValue(value, this::setValueAction);
    }

    /**
     * @return Get labels of all options
     */
    public final List<String> getOptions() {
        return getOptionsAction();
    }

    protected Select getSelector() {
        isSelector = true;
        return new Select(new Element(getLocator()).getWebElement());
    }

    protected List<WebElement> getElements() {
        if (!hasLocator() && allLabels() == null)
            throw exception("Can't check is element displayed or not. No optionsNamesLocator and allLabelsLocator found");
        if (allLabels() != null)
            try { return allLabels().avatar.searchAll().getElements(); }
            catch(Exception | Error ignore) { return new ArrayList<>(); }
        if (getLocator().toString().contains("%s"))
            throw exception("Can't check is element displayed or not. Please specify allLabelsLocator or correct optionsNamesLocator (should not contain '%s')");
        try { return getElementsFromTag(); }
        catch(Exception | Error ignore) { return new ArrayList<>(); }
    }

    public List<WebElement> getElementsFromTag() {
        List<WebElement> elements;
        try {
            elements = getAvatar().searchAll().getElements();
        } catch (Exception | Error ex) {
            return new ArrayList<>();
        }
        WebElement element = elements.get(0);
        if (elements.size() == 1)
            switch (element.getTagName()) {
                case "select":
                    return getSelector().getOptions();
                case "ul":
                    return element.findElements(By.tagName("li"));
            }
        return elements;
    }

    public WebElement getWebElement(String name) {
        return getElement(name).getWebElement();
    }

    protected boolean isDisplayedAction(String name) {
        WebElement element = getWebElement(name);
        return element != null && element.isDisplayed();
    }

    protected boolean isDisplayedAction(int num) {
        return isDisplayedInList(getElements(), num);
    }

    private boolean isDisplayedInList(List<WebElement> els, int num) {
        if (num <= 0)
            throw exception("Can't get option with num '%s'. Number should be 1 or more", num);
        if (els == null)
            throw exception("Can't find option with num '%s'. Please fix allLabelsLocator", num);
        if (els.size() < num)
            throw exception("Can't find option with num '%s'. Find '%s' options", num, els.size());
        return els.get(num - 1).isDisplayed();
    }

    protected boolean isDisplayedAction() {
        List<WebElement> els = avatar.findImmediately(this::getElements, null);
        return els != null && !els.isEmpty() && els.get(0).isDisplayed();
    }

    protected boolean waitDisplayedAction() {
        return timer().wait(() -> {
            List<WebElement> els = getElements();
            return els != null && !els.isEmpty() && els.get(0).isDisplayed();
        });
    }

    protected boolean waitVanishedAction() {
        return timer().wait(() -> !isDisplayedAction());
    }

    /**
     * @return Check is Element visible
     */
    public boolean isDisplayed() {
        return actions.isDisplayed(this::isDisplayedAction);
    }
    public boolean isDisplayed(String name) {
        return actions.isDisplayed(() -> isDisplayedAction(name));
    }
    public boolean isDisplayed(int num) {
        return actions.isDisplayed(() -> isDisplayedAction(num));
    }

    /**
     * @return Check is Element hidden
     */
    public boolean isHidden() {
        return actions.isDisplayed(() -> !isDisplayedAction());
    }

    /**
     * Waits while Element becomes visible
     */
    public void waitDisplayed() {
        actions.waitDisplayed(this::waitDisplayedAction);
    }

    /**
     * Waits while Element becomes invisible
     */
    public void waitVanished() {
        actions.waitVanished(() -> timer().wait(() -> !isDisplayedAction()));
    }
}